### PR TITLE
Balder/singlebit 2 bitvector

### PR DESCRIPTION
**Pull request by baldersheim**
@geirst PR
@toregge FYI

### CODE REVIEW

**Review by geirst — COMMENTED** [review]
Please have a look at my comments. I would also like to know which unit tests that instantiate and test this bool attribute (if any). We must ensure that all aspects of this is tested (etc. update / clear values, search, save and load).

**Comment by geirst** [comment]
@toregge please have a look as well

File: searchlib/src/vespa/searchlib/attribute/singleboolattribute.cpp [hunk 1]
```diff
@@ -0,0 +1,252 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/singleboolattribute.cpp]: Fixed

File: searchlib/src/vespa/searchlib/attribute/singleboolattribute.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+#pragma once
+
+#include "integerbase.h"
+#include <vespa/searchlib/common/growablebitvector.h>
+#include <limits>
+
+namespace search {
+
+class SingleBoolAttribute : public IntegerAttributeTemplate<int8_t>
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/singleboolattribute.h]: Fixed

**Comment by baldersheim** [comment]
It is using the existing tests that was there for SingleBItNumericAttribute.
Most of them are in attribute_test.cpp and in searchcontext.cpp.
It is not obvious in this PR as the type was changed to BOOL in an earlier PR.

File: searchlib/src/vespa/searchlib/attribute/singleboolattribute.h [hunk 3]
```diff
@@ -0,0 +1,102 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/singleboolattribute.h]: Fixed in #7996

File: searchlib/src/vespa/searchlib/attribute/singleboolattribute.h [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    void clearDocs(DocId lidLow, DocId lidLimit) override;
+    void onShrinkLidSpace() override;
+    void removeOldGenerations(generation_t firstUsed) override;
+    void onGenerationChange(generation_t generation) override;
+    uint64_t getEstimatedSaveByteSize() const override;
+
+    SearchContext::UP
+    getSearch(std::unique_ptr<QueryTermSimple> term, const attribute::SearchContextParams & params) const override;
+
+    uint32_t getValueCount(DocId doc) const override {
```
Comment (toregge) [inline searchlib/src/vespa/searchlib/attribute/singleboolattribute.h]: Please move implementation of most virtual method to the .cpp file.

Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/singleboolattribute.h]: This is a pattern used in all numeric variants. I think we should look at them in the same context.

**Pull request merged**